P4_14->alternate architecture conversion

ChrisDodd opened the pull request:
- initial work to split P4_14->16 conversion into a base converted with architecture specific subclass conversion, to allow for eventual P4_14->PSA conversion as well as target-specific architectures
- experimental support for extern functions with constant args returning extern instances (factory methods)

ChrisDodd commented:
Good idea -- I added a couple of simple test cases that revealed a parser problem and some type inferencing issues.

ChrisDodd commented:
So this change 'breaks' p4_16_errors/issue345.p4 -- it make it no longeer flag it as an error.  That happens because it reorders the order in which things get visited, and the error check that test is checking is order dependent.  Fixing the error check to work in either order breaks a bunch of other tests, amount them p4_16_samples/issue871.p4

The error check is giving an error when in type unification it tries to unify a type var with `int` from an untype integer literal.  This is a problem, but only if there isn't some other constraint that binds to a real type.  The problem with issue345.p4 is that while it has a type var that is used with an `int`, it is only a fragment with just the templated declaration and no instantiation.  So an error should /only/ occur if the template is later instantiated /without/ an explicit type.  issue871.p4 has a similar situation, but does have a template instantiation and that instantiation does have an explicit type, so it is (correctly) not an error.

For now I've just removed the issue345.p4 test -- I tried look at #345 but the program there has many other syntactic errors and I was unable to reconstruct anything that  showed the issue.

mbudiu-vmw commented:
Issue 345 should be an error until we have a notion of type constraints, where we can specify that H should be able to unify with int. This is similar to Java.
The language is designed so that each declaration can be type-checked separately. These are not templates like in C++, they are generics.

mbudiu-vmw commented on testdata/p4_16_samples/factory1.p4:
```diff
@@ -13,9 +13,16 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
+#include <core.p4>
```
I don't think you should delete the original file factory1.p4.

mbudiu-vmw replied:
I see, git things you actually renamed issue345, but in fact you have deleted it.
Nevermind this comment.

mbudiu-vmw commented on testdata/p4_16_samples/hashext.p4:
```diff
@@ -0,0 +1,38 @@
+limitations under the License.
+*/
+
+extern hash_function<O> {
```
Maybe this works better if you make it `hash_function<O, T>`?

ChrisDodd replied:
Unfortunately, that doesn't really make sense -- a hash algorithm should be polymorphic over different amounts/types of inputs -- a single instantiation of this extern should not be tied to a particular input and might in fact be used fo mulitple different inputs (though an architecture might restrict it dynamically to a single use per packet).

mbudiu-vmw replied:
Actually I think it does make sense: think hardware. You will need to allocate resources for the specific types and alignment of these values that you need to access. So you will need different instances for different hashes.
Why do you think it is a problem to require people to allocate an instance for each call?

ChrisDodd commented on testdata/p4_16_samples/hashext.p4:
```diff
@@ -0,0 +1,38 @@
+
+control test(inout hdrs hdr) {
+    apply {
+        hdr.crc = crc_poly(16w0x801a).hash(hdr.h1);
```
That should be checked by the evaluator, I think -- I'll add an error test case.

mbudiu-vmw replied:
Even if you modify the evaluator, MethodInstance::resolve is called in many places in the code.

ChrisDodd commented on frontends/p4/typeChecking/typeChecker.cpp:
```diff
@@ -192,6 +192,10 @@ TypeVariableSubstitution* TypeInference::unify(const IR::Node* errorPosition,
 
     TypeConstraints constraints(typeMap->getSubstitutions());
     constraints.addEqualityConstraint(destType, srcType);
+    if (auto itv = destType->to<IR::ITypeVar>())
```
That's not correct syntax -- controls can't contain method declarations -- but if they could, that's precisely what this change is trying to address: H needs to be bound by the constructor of c, not the call to x.  I'll add another test case that illustrates this with a normal extern rather than one that uses a factory method returning the extern.

ChrisDodd replied:
Hmm -- I think this case should be correct and illustrate the problem:

    extern ext<H> {
        ext(H v);
        void method<T>(H h, T t);
    }

    control c() {
        ext(16w0) x;
        apply {
            x.method(0, 8w0);
        }
    }

but it fails either with out without this change -- it fails to propagate the bound type for H from the constructor to the method call.  It fails even with an explicit `16w0` for the first arg to method!

mbudiu-vmw replied:
If this case fails it is a bug; we should add it to our test suite.
H should be unified with `bit<16>` and T with `bit<8>`.

mbudiu-vmw replied:
Actually this program fails because we do not infer type parameters from constructor arguments. If you modify it as follows: `ext<bit<16>>(16w0) x;` it works. The compiler requires all constructors for generics to specify explicitly type parameter values.

mbudiu-vmw replied:
We do that because in general constructor arguments may not be enough to infer all type parameters. So it is simpler to just ask users to write the type parameters explicitly. I think this is very consistent. It is not an artifact of the implementation traversal order.

ChrisDodd commented:
The problem with issue345 is that the current error check depends on the order in which things are hit in the traversal -- a minor change elsewhere changes the order of traversal and causes the test to fail.  I can't figure out a way of fixing the check that doesn't also cause issue871.p4 to fail.  If issue345 is supposed to be an error, then why is issue871 supposed to not be an error?  In both cases the code unifies a type var to an untyped constant (an InfInt) -- its just in the issue871 case it also unifies to a bit<32>

mbudiu-vmw commented:
The rule is that all type variables must be bound in order to resolve an instantiation.
In 345 they are not (int is treated like a type variable itself, with a constraint that it can be unified only with bit or int). In 781 all type variables are bound: I to `bit<32>` and T to `bit<8>`.

mbudiu-vmw commented:
BTW: this has nothing to do with the traversal order: we just add new type constraints and in the end all variables must be accounted for.

ChrisDodd commented:
Ok, I removed the qeustionable inferencing of type arguments and added several more test cases.

It would be nice to be able to infer extern type args from the constructor arguments, but that's something that could be addressed later.  We also will likely need to introduce some kind of type constraints to make this more flexible, if only implicit ones.  We already allow architecture-specific type constraints on things, they're just not explicit in the langauge.
